cmd/autoretrieve/bitswap/provider.go

```diff
@@ -0,0 +1,210 @@
+// each CID, it'll check if it's present in the blockstore. If it is, it will
+// respond with that block, and if it isn't, it'll start a retrieval and
+// register the block to be sent later.
+func (provider *Provider) ReceiveMessage(ctx context.Context, sender peer.ID, incoming message.BitSwapMessage) {
+
+	// For each item in the incoming wantlist, we either need to say whether we
+	// have the block or not, or send the actual block.
+
+entryLoop:
+	for _, entry := range incoming.Wantlist() {
+
+		// Only respond to WANT_HAVE and WANT_BLOCK
+		if entry.WantType != wantTypeHave && entry.WantType != wantTypeBlock {
+			continue
+		}
+
+		// First we check the local blockstore
+		block, err := provider.blockManager.Get(entry.Cid)
+
+		switch {
+		case err == nil:
+			// If we did have the block locally, just respond with that
+			provider.respondPositive(sender, entry, block)
+		default:
+			// Otherwise, we will need to ask for it...
+			if err := provider.retriever.Request(entry.Cid); err != nil {
+				// If request failed, it means there's no way we'll be able to
+				// get that block at the moment, so respond with DONT_HAVE
+				provider.respondNegative(sender, entry)
+				continue entryLoop
+			}
+
+			var queue chan blocks.Block
+
+			provider.blockQueuesLk.Lock()
+			switch entry.WantType {
+			case wantTypeHave:
+				queue = provider.blockQueues[sender].wantHaveChan
+			case wantTypeBlock:
+				queue = provider.blockQueues[sender].wantBlockChan
+			}
+			provider.blockQueuesLk.Unlock()
+
+			provider.blockManager.GetAwait(entry.Cid, queue)
+		}
+	}
+}
+
+func (provider *Provider) ReceiveError(err error) {
+	logger.Errorf("Error receiving bitswap message: %v", err)
+}
+
+func (provider *Provider) PeerConnected(peer peer.ID) {
+	blockQueue := BlockQueue{
+		wantHaveChan:  make(chan blocks.Block, 10),
+		wantBlockChan: make(chan blocks.Block, 10),
+	}
+
+	// When a peer connects, create a queue for it
+	provider.blockQueuesLk.Lock()
+	provider.blockQueues[peer] = blockQueue
+	provider.blockQueuesLk.Unlock()
+
+	// And start a send routine each for the WANT_BLOCKs and WANT_HAVEs
+
+	go func() {
+		for block := range blockQueue.wantBlockChan {
+			msg := message.New(false)
+			msg.AddBlock(block)
+			if err := provider.network.SendMessage(context.Background(), peer, msg); err != nil {
+				logger.Errorf("Could not send block %s to %s: %v", block.Cid(), peer, err)
+			}
+		}
+	}()
+
+	go func() {
+		for block := range blockQueue.wantHaveChan {
+			msg := message.New(false)
+			msg.AddHave(block.Cid())
+			if err := provider.network.SendMessage(context.Background(), peer, msg); err != nil {
+				logger.Errorf("Could not send HAVE %s to %s: %v", block.Cid(), peer, err)
+			}
+		}
+	}()
+}
+
+func (provider *Provider) PeerDisconnected(peer peer.ID) {
+	// On disconnect, destroy the peer's queue
+	provider.blockQueuesLk.Lock()
+
+	// TODO: need to clean up waiting blocks in block manager
```

Comment:
yeah, just closing the channels here will likely cause things to panic elsewhere

Reply:
using the peertaskqueue is probably the right way forward here, no need for us to reimplement all that tricky logic